### DISCUSSION TITLE
Inference with C++ interface under OpenMP

### DISCUSSION THREAD

**Discussion by KawaSakiHakuButsu** [deepmodeling/deepmd-kit#1924] [Q&A] [answered]
When I make inference with C++ interface **under OpenMP** in my code, there are different errors about memory for repeated tests as following.
e. g. 1
<img width="271" alt="error_01" src="https://user-images.githubusercontent.com/82147508/190907153-a72a8f22-c111-4cab-b758-2b9790c8bd54.png">
e. g. 2
<img width="269" alt="error_02" src="https://user-images.githubusercontent.com/82147508/190907209-300eb37a-e63a-483d-a7ef-f89b7978ac12.png">
e. g. 3
<img width="222" alt="error_03" src="https://user-images.githubusercontent.com/82147508/190907306-edad6033-9ac3-4005-a3ea-3feab783419b.png">
Interestingly, if I "**cout**" something around the inference function "compute", all errors disappear. Meanwhile, as long as the line with "dp.compute" is **deleted** together with some necessary changes,  thers's also no errors.
Specifically, I use inference **under OpenMP** in the work-sharing construct "for" like `#pragma omp parallel for`.

To clarify the connection between inference with C++ and above errors, I also test via the example of H2O. The code is put in the below
```
#include "deepmd/DeepPot.h"
#include <iostream>
#include <omp.h>

typedef double ValueType;

int main(){
  deepmd::DeepPot dp ("graph.pb");
  #pragma omp parallel for num_threads(16) schedule(static, 2)
  for (int i = 0 ; i < 36 ; ++i){
    std::vector<ValueType > coord = {1., 0., 0., 0., 0., 1.5, 1. ,0. ,3.};
    std::vector<ValueType > cell = {10., 0., 0., 0., 10., 0., 0., 0., 10.};
    std::vector<int > atype = {1, 0, 1};
    double e;
    std::vector<ValueType > f, v;
    dp.compute(e, f, v, coord, atype, cell);
    std::cout << e << " " << omp_get_thread_num() << " " << omp_get_num_threads() << std::endl;
  }
```
However, sometimes there is error about (array) index out of bounds, like `Segmentation fault (core dumped)`.

So how can I use inference with C++ interface under OpenMP correctly?
Thank you!

**Comment by njzjz** [comment, marked as answer]
I don't think it's thread-safe. Also note that TensorFlow itself utilizes multiple threads.

**Reply by KawaSakiHakuButsu** [reply]
Could you provide more detailed explanation for the lack of thread-safety?
Thank you!

**Reply by njzjz** [reply]
I am unsure whether TensorFlow is thread-safe, but at least it never claims or is designed to be.

**Reply by KawaSakiHakuButsu** [reply]
I get your point now. 
However, when I want to do inference on different structures simultaneously, how can I improve calculation efficiency? If it's convenient for you, could you provide me with some ideas?
Thank you!

**Reply by njzjz** [reply]
I don't know that you program is but you can try MPI. If they are not required to infer in the same time, you do not need to run only a single program.

**Reply by KawaSakiHakuButsu** [reply]
Thank you for your suggestions, I'll have a try.
